api/v1alpha1/gateway_types.go

```diff
@@ -54,78 +57,129 @@ type GatewayList struct {
 }
 
+// ProtocolType defines the application protocol accepted by a Listener.
+type ProtocolType string
+
 const (
-	// HTTPProtocol constant.
-	HTTPProtocol = "HTTP"
-	// HTTPSProtocol constant.
-	HTTPSProtocol = "HTTPS"
+	// HTTPProtocolType accepts cleartext HTTP sessions.
+	HTTPProtocolType ProtocolType = "HTTP"
+
+	// HTTPSProtocolType accepts HTTPS sessions.
+	HTTPSProtocolType ProtocolType = "HTTPS"
+
+	// TLSProtocolType accepts TLS sessions over TCP.
+	TLSProtocolType ProtocolType = "TLS"
+
+	// TLSProtocolType accepts TCP sessions.
+	TCPProtocolType ProtocolType = "TCP"
 )
 
-// Listener defines a
+// Listener embodies the concept of a logical endpoint where a
+// Gateway can accept network connections.
+//
+// The GatewayClass may collapse compatible Listener definitions
+// into single implementation-defined acceptor.
+//
+// Listeners are considered compatible if the Protocol field is
+// "HTTP", HTTPS or "TLS", the Hostname field is specified, and the
+// Port field matches.
 type Listener struct {
-	// Name is the listener's name and should be specified as an
-	// RFC 1035 DNS_LABEL [1]:
+	// Hostname is the fully qualified domain name of a network host,
+	// as defined by [RFC 3986](https://tools.ietf.org/html/rfc3986#section-3.2.2).
+	// Note the following deviations from the "host" part of the URI as defined in the RFC:
 	//
-	// [1] https://tools.ietf.org/html/rfc1035
+	// 1. IPs are not allowed.
+	// 2. The `:` delimiter is not respected because ports are not allowed.
 	//
-	// Each listener of a Gateway must have a unique name. Name is used
-	// for associating a listener in Gateway status.
+	// Hostname specifies the virtual hostname for protocol
+	// types that define this concept. If the protocol is "HTTP"
+	// or "HTTPS", Hostname is must be present and must not be an
+	// empty string.
 	//
-	// Support: Core
+	// Incoming requests are matched against Hostname before processing HTTPRoute
+	// rules. For example, if the request header contains host: foo.example.com,
+	// an HTTPRoute with hostname foo.example.com will match. However, an
+	// HTTPRoute with hostname example.com or bar.example.com will not match.
+	// If Hostname is unspecified, the Gateway routes all traffic based on
+	// the specified rules.
 	//
-	// +required
-	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
-	// Address requested for this listener. This is optional and behavior
-	// can depend on GatewayClass. If a value is set in the spec and
-	// the request address is invalid, the GatewayClass MUST indicate
-	// this in the associated entry in GatewayStatus.Listeners.
+	// Hostname must be supplied required for "HTTP" and "HTTPS" protocols, and may
+	// be supplied for "TLS" protocol.
 	//
-	// Support:
+	// Support: Core
 	//
 	// +optional
-	Address *ListenerAddress `json:"address,omitempty" protobuf:"bytes,2,opt,name=address"`
-	// Port is a list of ports associated with the Address.
+	// +kubebuilder:validation:MinLength=1
```

Comment:
Are these two concepts compatible?

Reply:
Maybe? What I'm trying to say is that the field is optional, but must not be the empty string.

Reply:
For strings, I think the common pattern would be to use a pointer here to indicate that it was optional. That way nil is valid and empty string is not. Since pointers can be a pain to work with, a better option might be to just document that an empty string will be ignored/treated as unspecified. I think that would mean removing the kubebuilder validation and instead relying on implementations to ignore empty string values here.

Reply:
The hostname is now an optional struct, not a string.